[JIT] Enable freezing for sparse COO tensors

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69614

Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

Differential Revision: [D32954620](https://our.internmc.facebook.com/intern/diff/D32954620)

davidberard98 committed:
[JIT] Enable freezing for sparse COO tensors

Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Enable freezing for sparse COO tensors"


Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

davidberard98 committed:
Update on "[JIT] Enable freezing for sparse COO tensors"


Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

Differential Revision: [D32954620](https://our.internmc.facebook.com/intern/diff/D32954620)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

davidberard98 committed:
Update on "[JIT] Enable freezing for sparse COO tensors"


Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

Differential Revision: [D32954620](https://our.internmc.facebook.com/intern/diff/D32954620)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

eellison commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -291,16 +309,7 @@ struct TORCH_API IValue final {
-      }
-
-      return thisTensor.is_alias_of(rhsTensor);
+      return isAliasOf(this->toTensor(), rhs.toTensor());
```
nice refactoring!

eellison commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -872,17 +881,22 @@ struct TORCH_API IValue final {
+        // to use mkldnn storage
+        return reinterpret_cast<size_t>(ten.unsafeGetTensorImpl());
+      } else if (ten.is_sparse()) {
+        return hashTensor(ten._values());
```
so, our options are:
- hash on just index/values
this makes it so we will detect aliasing overlap between sparse tensors and non-sparse tensors only of one of the two.  
- maintain separate hashes for index and value... i think this is probably not worth it for what is the extremely rare (non-existant ?) case of mutating after freezing sparse tensors indices that also happens to be a tensor that is used elsewhere
- 
i think current approach is good, maybe we can  add comment re-ignoring indices

eellison commented on test/jit/test_sparse.py:
```diff
@@ -0,0 +1,39 @@
+            def forward(self, x):
+                return x + self.a
+
+        torch.jit.save(torch.jit.script(SparseTensorModule()), io.BytesIO())
```
can we test load as well? there are examples in test_jit of how to do that

eellison commented on test/jit/test_sparse.py:
```diff
@@ -0,0 +1,39 @@
+        m.eval()
+        frozen = torch.jit.freeze(torch.jit.script(m))
+
+        frozen_result = frozen.forward(x)
```
can we test save-load here just to be safe ?

eellison reviewed (approved):
Nice!! couple of small comments

davidberard98 committed:
Update on "[JIT] Enable freezing for sparse COO tensors"


Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

Differential Revision: [D32954620](https://our.internmc.facebook.com/intern/diff/D32954620)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Enable freezing for sparse COO tensors"


Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

Differential Revision: [D32954620](https://our.internmc.facebook.com/intern/diff/D32954620)

[ghstack-poisoned]

davidberard98 commented:
Turns out there was still an issue with freezing/serializing modules that had more than 1 sparse tensor, updated to fix that

davidberard98 committed:
Update on "[JIT] Enable freezing for sparse COO tensors"


Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

Differential Revision: [D32954620](https://our.internmc.facebook.com/intern/diff/D32954620)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

davidberard98 committed:
Update on "[JIT] Enable freezing for sparse COO tensors"


Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

Differential Revision: [D32954620](https://our.internmc.facebook.com/intern/diff/D32954620)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).

davidberard98 committed:
Update on "[JIT] Enable freezing for sparse COO tensors"


Previously sparse COO tensors were ignored during freezing, because
`tryInsertConstant` would fail during `freeze_module.cpp`, and because
hashes weren't implemented for COO tensor IValues.

Differential Revision: [D32954620](https://our.internmc.facebook.com/intern/diff/D32954620)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32954620).
